Feature/region view pip table

amkwong opened the pull request:
## Ticket
N/A

## Purpose
- Added a table in region view
- Added the PIP cluster column to the table in both single-variant and region views
- Changed hover tooltips in dropdown menu buttons to popovers
- Made some changes to wording in some popover text
- Updated region view UCSC link to highlight the region being viewed
- Added a button on the top of region view to go to the table
- Added css to prevent line breaks in dropdown menu items
- Updates from #55 have been merged to this branch, because the PIP cluster updates require the display fix for PIP values from there (as a consequence, #55 has been closed)

## How to test this
- Check both single-variant and region views to make sure table is rendering correctly
- Click buttons at the top navbar to make sure they bring you to the correct places
- Check dropdown menus to make sure there are no errant line breaks in button descriptions

## Deployment / configuration notes
(none)

abought commented on pheget/api/__init__.py:
```diff
@@ -129,3 +129,24 @@ def variant_query(chrom: str, pos: int):
+
+@api_blueprint.route(
+    "/region/<string:chrom>/<int:start>-<int:end>/data/", methods=["GET"]
+)
```
Can you help me understand the difference in requirements between this endpoint and `region_query` above, in the same file? I may be missing an important subtlety, and there are no unit tests to guide on the differences.

It seems they are overall very similar, except with minor differences in optional query params (tissue vs piponly), API response formatting (`{}` vs `{'data': {}`), and query options (`chr` prefix handling).

Overall it sort of leaves the impression that both endpoints *could* be consolidated into one. From a maintainability/API standpoint, it can be very jarring if every view in your app obeys different query syntax and formatting rules.

amkwong replied:
You're correct, this was just a quick hack to test ways to populate the table. I thought one way to consolidate this is to add an option in the original region query API so that we can request data only vs. returning {"data": data}, but using the Tabulator ajaxResponse callback looks to be cleaner.

abought commented on pheget/api/format.py:
```diff
@@ -328,6 +329,10 @@ def query_variants(
 
+    if piponly:
+        reader.add_filter("pip")
+        reader.add_filter(lambda result: result.pip > 0.0)
```
Is there any special reason why this cutoff was chosen? What is the difference between a variant with no pip data, and a variant with a pip value of 0?

In cases where code depends on hardcoded constants (like pip >0), it can be helpful to add a small comment describing how that constant was chosen.

amkwong replied:
Because the way we currently implement appending PIP data to data points sets the default value to 0.0, this filter is just a crude way to filter the data down to only points that have non-missing PIP data - in the PIP database no points have a value of 0.

I will add a comment to note this.

abought commented:
Incidentally, per the difference in API response format (`{ data: []}` vs `[]`), there is an additional tabulator option that may be of use:
http://tabulator.info/docs/4.6/data#ajax-response

amkwong commented on src/views/Region.vue:
```diff
@@ -30,6 +33,12 @@ function getData(queryParams) {
     .then((resp) => resp.json());
 }
 
+// function getTopPipData(chrom, start, end) {
```
Yes, this has been removed.

amkwong commented:
> Incidentally, per the difference in API response format (`{ data: []}` vs `[]`), there is an additional tabulator option that may be of use:
> http://tabulator.info/docs/4.6/data#ajax-response

I incorporated changes to use this callback and removed the redundant API endpoint in 7bc3e74.

amkwong commented:
Updates in 042f39d: I added the piponly option to the range API (?piponly=True will return only points with non-missing PIP values), and added an extra comment for its potential future use - If large regions sending too much data becomes a problem, then we can use this filter as a first attempted fix.

abought commented:
>  If large regions sending too much data becomes a problem, then we can use this filter as a first attempted fix.

Overall it looks good for now and is ok to merge. Thanks for helping with the questions and I'm glad the tabulator option worked for you.

It's hard to comment on the strategy for limiting query size/time, because the following variables are hard for me to assess on my laptop:

1. How much data (unfiltered) is involved in a typical real query for this use case? If the tabix query + parsing is excessive, we'll still have performance issues. In that case, a long term solution might prefer to do the filtering in, eg, a database (reduce the amount of data that python needs to act on)
2. What fraction of data has pip=0? (eg, what % of the data is eliminated by using this filter?)
3. Does the PIP have a specific meaning that would lend itself to a more meaningful significance threshold than "present or absent"? (this could tie into the "top hits" concept of the original feature proposal)

amkwong commented:
> 1. How much data (unfiltered) is involved in a typical real query for this use case? If the tabix query + parsing is excessive, we'll still have performance issues. In that case, a long term solution might prefer to do the filtering in, eg, a database (reduce the amount of data that python needs to act on)

This depends on the size of the region, but for a large (1mbp) dense region this can run into hundreds of thousands and possibly millions of data points. A database is probably a good option anyway, once I have some time to work on this.

> 2. What fraction of data has pip=0? (eg, what % of the data is eliminated by using this filter?)

The DAPG database contains 21 million data points; the full GTEx v8 database contains somewhere in the range of 10 million points per tissue across 49 tissues, with some missing data, so a little under 500 million data points. Filtering by PIP should roughly remove 95% of the points, give or take.

> 3. Does the PIP have a specific meaning that would lend itself to a more meaningful significance threshold than "present or absent"? (this could tie into the "top hits" concept of the original feature proposal)

Since PIP is a Bayesian posterior probability, the interpretation of whether the PIP value can be meaningfully parsed depends on how much you believe the generating model. Trying to set any kind of threshold for "significance" may be an impossible task (though some variants are obviously more important than others). 

That said, if a variant was picked up by DAP-G to be included, then it means there's at least some evidence for the variant being important (or at least in LD with something that's important), and that there's a chance it'll be interesting to a researcher out there somewhere. This is the reason I went with the "present in the PIP database" criterion for the filter.

abought commented:
Thanks for the thoughtful and informative answers; very helpful!

Based on these comments, it sounds like this will help with payload size. It also sounds like an improved backend could yield further gains (by removing 95% of the parsing time spent looking at rows of interest).
